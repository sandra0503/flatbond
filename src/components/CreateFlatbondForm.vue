<template>
  <b-form class="create-flatbond-form" @submit.prevent="onSubmit">
    <b-form-group
      :invalid-feedback="invalidFeedbackRent"
      :state="$v.form.rent.$dirty ? !$v.form.rent.$error : null"
      label="Rent:"
      label-for="rent"
    >
      <b-input-group prepend="£" append=".00">
        <b-form-input
          :state="$v.form.rent.$dirty ? !$v.form.rent.$error : null"
          :number="true"
          v-model="form.rent"
          placeholder="Your rent"
          id="rent-input"
          trim
          @input="setRent"
        ></b-form-input>
      </b-input-group>
    </b-form-group>

    <div>
      <b-form-radio
        v-model="form.rentPeriod"
        name="payment-period-radios"
        value="weekly"
        >Weekly</b-form-radio
      >
      <b-form-radio
        v-model="form.rentPeriod"
        name="payment-period-radios"
        value="monthly"
        >Monthly</b-form-radio
      >
    </div>

    <b-form-group
      :invalid-feedback="invalidFeedbackPostcode"
      :state="$v.form.postcode.$dirty ? !$v.form.postcode.$error : null"
      class="fieldset-postcode"
      label="Postcode:"
      label-for="postcode"
    >
      <b-form-input
        :state="$v.form.postcode.$dirty ? !$v.form.postcode.$error : null"
        v-model="form.postcode"
        id="postcode-input"
        placeholder="Postcode"
        trim
        @input="setPostcode"
      ></b-form-input>
    </b-form-group>

    <b-form-group label="Your one-off membership fee:" label-for="fee">
      <div class="membership-fee">£ {{ membershipFee }}</div>
    </b-form-group>

    <div class="error" v-if="error">
      Ooops. Something went wrong. Please check your data and try again.
    </div>
    <b-button
      :disabled="$v.form.$invalid"
      type="submit"
      class="submitButton"
      variant="primary"
      >Submit</b-button
    >
  </b-form>
</template>

<script>
import { mapState } from "vuex";

import {
  between,
  minLength,
  numeric,
  required
} from "vuelidate/lib/validators";

import {
  MINIMUM_MONTHLY,
  MAXIMUM_MONTHLY,
  MINIMUM_WEEKLY,
  MAXIMUM_WEEKLY,
  VAT_FACTOR,
  WEEKLY_FACTOR,
  MINIMUM_MEMBERSHIP_FEE
} from "../config.json";

import {
  BButton,
  BForm,
  BFormGroup,
  BFormInput,
  BFormRadio,
  BInputGroup
} from "bootstrap-vue";

const PERIODS = {
  MONTHLY: "monthly",
  WEEKLY: "weekly"
};

export default {
  name: "CreateFlatbondForm",
  components: {
    "b-button": BButton,
    "b-form": BForm,
    "b-form-group": BFormGroup,
    "b-form-input": BFormInput,
    "b-form-radio": BFormRadio,
    "b-input-group": BInputGroup
  },
  data() {
    return {
      form: {
        rent: null,
        rentPeriod: PERIODS.WEEKLY,
        postcode: null
      },
      error: false
    };
  },
  computed: {
    ...mapState(["feeConfig"]),
    minimumRent() {
      return this.form.rentPeriod === PERIODS.MONTHLY
        ? MINIMUM_MONTHLY
        : MINIMUM_WEEKLY;
    },
    maximumRent() {
      return this.form.rentPeriod === PERIODS.MONTHLY
        ? MAXIMUM_MONTHLY
        : MAXIMUM_WEEKLY;
    },
    invalidFeedbackRent() {
      if (this.form.rentPeriod === PERIODS.MONTHLY) {
        return `Your monthly rent should be a number between ${MINIMUM_MONTHLY} and ${MAXIMUM_MONTHLY}.`;
      }
      return `Your weekly rent should be a number between ${MINIMUM_WEEKLY} and ${MAXIMUM_WEEKLY}.`;
    },
    invalidFeedbackPostcode() {
      return "Please enter a valid postcode.";
    },
    weeklyRent() {
      if (!Number.isInteger(parseFloat(this.form.rent))) {
        return null;
      }

      return this.form.rentPeriod === PERIODS.WEEKLY
        ? this.form.rent
        : this.form.rent / WEEKLY_FACTOR;
    },
    monthlyRentInPence() {
      if (!this.weeklyRent) {
        return null;
      }
      return this.form.rentPeriod === PERIODS.WEEKLY
        ? (this.weeklyRent * WEEKLY_FACTOR * 100).toFixed()
        : (this.form.rent * 100).toFixed();
    },
    membershipFee() {
      if (this.weeklyRent && this.feeConfig.fixedMembershipFee) {
        return this.fixedFee ? this.fixedFee : null;
      }

      return this.flexibleFee;
    },
    fixedFee() {
      const fee =
        this.feeConfig.fixedMembershipFeeAmount &&
        this.feeConfig.fixedMembershipFeeAmount / 100;
      return this.getRoundNumber(this.addVAT(fee));
    },
    flexibleFee() {
      if (!this.weeklyRent) {
        return 0;
      }

      const feeBasis = Math.max(MINIMUM_MEMBERSHIP_FEE, this.weeklyRent);

      return this.getRoundNumber(this.addVAT(feeBasis));
    }
  },
  validations() {
    return {
      form: {
        rent: {
          required,
          numeric,
          between: between(this.minimumRent, this.maximumRent)
        },
        postcode: {
          required,
          minLength: minLength(3)
        }
      }
    };
  },
  methods: {
    onSubmit() {
      this.$store
        .dispatch("CREATE_FLATBOND", {
          rent: this.monthlyRentInPence,
          postcode: this.form.postcode
        })
        .then(() => {
          this.$store.dispatch("SET_FEE", this.membershipFee);
          this.$router.push("/success");
        })
        .catch(() => {
          this.error = true;
        });
    },
    setRent(rentValue) {
      this.rent = rentValue;
      this.$v.form.rent.$model = this.rent;
      this.resetErrorMessage();
    },
    setPostcode(postcode) {
      this.postcode = postcode.toUpperCase();
      this.$v.form.postcode.$model = this.postcode;
      this.resetErrorMessage();
    },
    resetErrorMessage() {
      this.error = false;
    },
    addVAT(amount) {
      return amount * VAT_FACTOR;
    },
    getRoundNumber(number) {
      return parseInt(number.toFixed());
    }
  }
};
</script>

<style scoped lang="scss">
.create-flatbond-form {
  text-align: left;
  max-width: 480px;
  margin: auto;
}

.error {
  color: var(--red);
}

.btn {
  margin: 25px auto;
  display: block;
}

.membership-fee {
  font-family: "Lato", sans-serif;
  font-weight: 700;
  background: #f3f5f6;
  border-radius: 0.25rem;
  text-align: center;
  padding: 10px 50px;
}
</style>
