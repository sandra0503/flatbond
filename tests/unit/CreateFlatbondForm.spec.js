import { shallowMount } from "@vue/test-utils";
import CreateFlatbondForm from "@/components/CreateFlatbondForm";

describe("CreateFlatbondForm", () => {
  const build = () => {
    const wrapper = shallowMount(CreateFlatbondForm);

    return {
      wrapper
    };
  };

  it("renders the component and matches snapshot", () => {
    const { wrapper } = build();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
