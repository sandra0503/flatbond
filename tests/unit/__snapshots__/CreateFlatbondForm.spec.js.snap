// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateFlatbondForm renders the component and matches snapshot 1`] = `
<form class="create-flatbond-form">
  <div role="group" class="form-group"><label for="rent" class="d-block">Rent:</label>
    <div>
      <div role="group" class="input-group">
        <div class="input-group-prepend">
          <div class="input-group-text">£</div>
          <!---->
        </div><input id="rent-input" type="text" placeholder="Your rent" class="form-control">
        <div class="input-group-append">
          <div class="input-group-text">.00</div>
          <!---->
        </div>
      </div>
      <div tabindex="-1" role="alert" aria-live="assertive" aria-atomic="true" class="invalid-feedback">Your weekly rent should be a number between 25 and 2000.</div>
      <!---->
      <!---->
    </div>
  </div>
  <div>
    <div class="custom-control custom-radio"><input type="radio" name="payment-period-radios" autocomplete="off" class="custom-control-input" value="weekly"><label class="custom-control-label">Weekly</label></div>
    <div class="custom-control custom-radio"><input type="radio" name="payment-period-radios" autocomplete="off" class="custom-control-input" value="monthly"><label class="custom-control-label">Monthly</label></div>
  </div>
  <div role="group" class="form-group fieldset-postcode"><label for="postcode" class="d-block">Postcode:</label>
    <div><input id="postcode-input" type="text" placeholder="Postcode" class="form-control">
      <div tabindex="-1" role="alert" aria-live="assertive" aria-atomic="true" class="invalid-feedback">Please enter a valid postcode.</div>
      <!---->
      <!---->
    </div>
  </div>
  <div role="group" class="form-group"><label for="fee" class="d-block">Your one-off membership fee:</label>
    <div>
      <div class="membership-fee">£ 0</div>
      <!---->
      <!---->
      <!---->
    </div>
  </div>
  <!----> <button type="submit" disabled="disabled" class="btn submitButton btn-primary disabled">Submit</button>
</form>
`;
